src/mesh/mesh.cpp

```diff
@@ -1184,6 +1173,68 @@ void Mesh::DoStaticRefinement(ParameterInput *pin) {
     pib = pib->pnext;
   }
 }
+
+void Mesh::BuildAndRegisterCommBuffers_() {
+  // Clear boundary communication buffers
+  boundary_comm_map.clear();
+  pcoalesced_comms->clear();
+
+  // JMM: Estimates about chunk size, pre-allocation, etc., are tuned to
+  // all blocks on a rank, so we use the base partition.
+
+  // TODO(JMM): I don't like that we need to add base with the base
+  // partition here. We shouldn't have to, but I don't know how to fix
+  // it without a refactor.
+  BuildBoundaryBuffers(mesh_data.Add("base", GetBasePartition()));
+  if (do_coalesced_comms) RegisterCoalescedComms(this);
+
+  if (multigrid) { // But... multigrid is sufficiently hairy that I'm
+                   // going to let LFR figure this one out later.
+    for (int gmg_level = GetGMGMinLevel(); gmg_level <= GetGMGMaxLevel(); ++gmg_level) {
+      const auto grid_id = GridIdentifier::two_level_composite(gmg_level);
+      for (auto &partition : GetDefaultBlockPartitions(grid_id)) {
+        auto &mdg = mesh_data.Add("base", partition);
+        BuildBoundaryBuffers(mdg);
+        BuildGMGBoundaryBuffers(mdg);
+        if (do_coalesced_comms) RegisterCoalescedCommsGMG(mdg);
+      }
+    }
+  }
+
+  if (do_coalesced_comms) {
+    pcoalesced_comms->ResolveAndSendSendBuffers();
+    // This operation is blocking
+    pcoalesced_comms->ReceiveBufferInfo();
+  }
+}
+
+bool Mesh::TryReallocCommBufferPools() {
+  bool realloc = false;
+  for (auto &[k, pool] : pool_map) {
+    std::size_t inuse = pool.NumBuffersInUse();
+    std::size_t total = pool.NumBuffersInPool();
+    std::size_t delta = total - inuse;
+    if ((inuse < buffer_reset_frac_ * total) && (delta > CommBufferChunkSize())) {
+      realloc = true;
+      break;
+    }
+  }
+  if (realloc) {
+    // The buffer pool must be cleared out, since otherwise, buffers
+    // will just be reference-count-freed
+    for (auto &[k, pool] : pool_map) {
+      pool.Clear();
+    }
+    // We need to clear the caches because comm buffers are kokkos
+    // views, which are reference counted, and they're stored in the
+    // caches.
```

Comment:
`BvarsSubCache_t::buf_vec` is a `std::vector<CommBuffer<buf_pool_t<Real>::owner_t> *>`, so we need to clear the caches because these pointers are no longer valid.